tests/typecheck-tests.ts

```diff
@@ -542,7 +542,72 @@ import { ok, err, okAsync, errAsync, Result, ResultAsync } from '../src'
 
 
 (function describe(_ = 'Combine on Unbounded lists') {
-  // TODO:
-  // https://github.com/supermacro/neverthrow/issues/226
+  (function describe(_ = 'combine') {
+    (function it(_ = 'combines different results into one') {
+      type Expectation = Result<[ number, string ], Error | string[]>;
```

Comment:
This expectation is not correct - perhaps there may some miscommunication / misunderstanding of the desired goal of #226. 

The desired outcome when fixing #226 is that all the `combine` functions can correctly produce a **tuple** whose length / size is equal to the length / size of the array passed into the `combine` functions.

In other words, we need to treat the list argument to the `combine` functions as a tuple instead of a variable-length list.

Example:

When calling:

```typescript
Result.combine([
  ok(1),
  ok('string'),
  err([ 'string', 'string2' ]),
  err(new Error('error content')),
])
```

The `resultList` argument should understand that we have a fixed-length tuple whose type is:

```typescript
[ Result<number, never>, Result<string, never>, Result<never, string>, Result<never, Error> ]
```

Therefore, the `Expectation` of this test would be that we should have:

```typescript
Result<[ number, string, never, never ], Error | string[] | never>;
```

---

Additionally - in practice, `combine` would not be called with the `ok` and `err` constructors (or the `okAsync` / `errAsync` constructors), rather, we'd have a list of `Result<T, E>` values where the `T`s and the `E`s are not `unknown`.

So to make these tests a bit more realistic, I would provide explicit type arguments to the `ok` and `err` function invocations like this:

```typescript
type Expectation = Result<[ number, string, boolean ], Error | string[]>;

const result: Expectation = Result.combine([
  ok<number, string>(1),
  ok<string, string>('string'),
  err<boolean, string[]>([ 'string', 'string2' ]),
  err<boolean, Error>(new Error('error content')),
])
```

Reply:
Ok as the types did already pass the tests in the examples that I have not added, I can still inspect the ones that you mention. Add your expectations and I'll see them. After this point, obviously, it will not be hard to do the rest. @supermacro

Reply:
Hello @supermacro, I have improved the typing system and made some of the lighter for CPU, and also easier for type matching. Yet, I have not touched anything else mentioned other than this conversation page.